maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java

```diff
@@ -0,0 +1,245 @@
+  private final Map<MavenProjectKey, Span> mavenProjectKeySpanMap = new HashMap<>();
+  private Span rootSpan;
+
+  @Nullable
+  public Span getRootSpan() {
+    return rootSpan;
+  }
+
+  /** @throws IllegalStateException Root span already defined */
+  public void setRootSpan(@Nonnull Span rootSpan) throws IllegalStateException {
+    if (this.rootSpan != null) {
+      throw new IllegalStateException("Root span already defined " + this.rootSpan);
+    }
+    this.rootSpan = rootSpan;
+  }
+
+  @Nonnull
+  public Span getSpan(@Nonnull MavenProject mavenProject) {
+    final MavenProjectKey key = MavenProjectKey.fromMavenProject(mavenProject);
+    final Span span = this.mavenProjectKeySpanMap.get(key);
+    if (span == null) {
+      throw new IllegalStateException(
+          "Span not started for project "
+              + mavenProject.getGroupId()
+              + ":"
+              + mavenProject.getArtifactId());
+    }
+    return span;
+  }
+
+  @Nonnull
+  public Span getRootSpanNotNull() {
+    if (rootSpan == null) {
+      throw new IllegalStateException("Root span not defined");
+    }
+    return rootSpan;
+  }
+
+  @Nonnull
+  public Span removeRootSpan() {
+    if (rootSpan == null) {
+      throw new IllegalStateException("Root span not defined");
+    }
+    if (!this.mojoExecutionKeySpanMap.isEmpty()) {
+      throw new IllegalStateException(
+          "Remaining children spans: "
+              + this.mojoExecutionKeySpanMap.keySet().stream()
+                  .map(MojoExecutionKey::toString)
+                  .collect(Collectors.joining(", ")));
+    }
+    return rootSpan;
+  }
+
+  public void putSpan(@Nonnull Span span, @Nonnull MavenProject mavenProject) {
+    MavenProjectKey key = MavenProjectKey.fromMavenProject(mavenProject);
+    Span previousSpanForKey = mavenProjectKeySpanMap.put(key, span);
+    if (previousSpanForKey != null) {
+      throw new IllegalStateException();
+    }
+  }
+
+  @Nonnull
+  public Span removeSpan(@Nonnull MavenProject mavenProject) throws IllegalStateException {
+    MavenProjectKey key = MavenProjectKey.fromMavenProject(mavenProject);
+    Span span = mavenProjectKeySpanMap.remove(key);
+    if (span == null) {
+      throw new IllegalStateException();
+    }
+    return span;
+  }
+
+  public void putSpan(@Nonnull Span span, @Nonnull MojoExecution mojoExecution) {
+    MojoExecutionKey key = MojoExecutionKey.fromMojoExecution(mojoExecution);
+    Span previousSpanForKey = mojoExecutionKeySpanMap.put(key, span);
+    if (previousSpanForKey != null) {
+      throw new IllegalStateException();
+    }
+  }
+
+  @Nonnull
+  public Span removeSpan(@Nonnull MojoExecution mojoExecution) throws IllegalStateException {
+    MojoExecutionKey key = MojoExecutionKey.fromMojoExecution(mojoExecution);
+    Span span = mojoExecutionKeySpanMap.remove(key);
+    if (span == null) {
+      throw new IllegalStateException();
+    }
+    return span;
+  }
+
+  private static class MavenProjectKey {
```

Comment:
We can use auto-value for these

https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/aws-xray/build.gradle.kts#L27
https://github.com/open-telemetry/opentelemetry-java-contrib/blob/ed31fbaac9b936e0b68152f3c74412ef8c8d08c2/aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/GetSamplingTargetsRequest.java#L13